mrmustard/physics/gaussian_integrals.py

```diff
@@ -472,3 +392,112 @@ def contract_two_Abc_poly(
+
+    The functions are assumed to be in the form
+    :math:`F_i(z) = c_i\textrm{exp}(0.5 z^T A_i z + b_i^T z),\quad z\in\mathbb{C}^{n_i}`
+
+    where ``A_i`` is complex symmetric
+
+    The integration measure is given by (``s`` here is the ``measure`` argument):
+    :math:`dmu(z) = \textrm{exp}(s * |z|^2) \frac{d^{2n}z}{\pi^n} = \frac{1}{\pi^n}\textrm{exp}(s * |z|^2) d\textrm{Re}(z) d\textrm{Im}(z)`
+
+    Arguments:
+        A1,b1,c1: the ``(A,b,c)`` triple that defines ``F_1(z)``
+        A2,b2,c2: the ``(A,b,c)`` triple that defines ``F_2(z^*)``
+        idx1: the tuple of indices indicating which variables of F_1 to integrate over
+        idx2: the tuple of indices indicating which variables of F_2 to integrate over
+        measure: the exponent of the measure (default is -1: Bargmann measure)
+
+    Returns:
+        The ``(A,b,c)`` triple of the result of the integral.
+
+    Raises:
+        ValueError: If ``idx1`` and ``idx2`` have different lengths or if A_i, b_i and c_i have non-matching batch size.
+    """
+    # first we ensure that Ai, bi, ci have the correct shapes
+    A1, b1, c1 = Abc1
+    A2, b2, c2 = Abc2
+    c1 = math.astensor(c1)
+    c2 = math.astensor(c2)
+    if not batched:
+        A1 = math.atleast_3d(A1)
+        A2 = math.atleast_3d(A2)
+        b1 = math.atleast_2d(b1)
+        b2 = math.atleast_2d(b2)
+        c1 = math.atleast_1d(c1)
+        c2 = math.atleast_1d(c2)
+
+    if len(idx1) != len(idx2):
+        raise ValueError(
+            f"idx1 and idx2 must have the same length, got {len(idx1)} and {len(idx2)}"
+        )
+    if not A1.shape[0] == b1.shape[0] == c1.shape[0] == A2.shape[0] == b2.shape[0] == c2.shape[0]:
+        raise ValueError(
+            f"Batch size mismatch: got {A1.shape[0]} for A1, {b1.shape[0]} for b1, {c1.shape[0]} for c1, {A2.shape[0]} for A2, {b2.shape[0]} for b2, {c2.shape[0]} for c2."
+        )
+
+    # get various sizes for later
+    (batch, n1, _) = A1.shape
+    (_, n2, _) = A2.shape
+    d1 = c1.shape[1:]
+    d2 = c2.shape[1:]
+    m = len(idx1)  # number of variable pairs to integrate over
+    idx = tuple(idx1) + tuple(i + n1 for i in idx2)
+
+    A = np.zeros((batch, n1 + n2, n1 + n2), dtype=A1.dtype)
+    for i in range(batch):
+        A[i] = math.block_diag(A1[i], A2[i])
+    b = math.concat([b1, b2], axis=-1)
+    c = math.einsum("bj,bk->bjk", math.reshape(c1, (batch, -1)), math.reshape(c2, (batch, -1)))
+    c = math.reshape(c, (batch,) + d1 + d2)
+
+    if len(idx) == 0:
+        return A, b, c
+
+    not_idx = tuple(i for i in range(n1 + n2) if i not in idx)
+
+    I = math.eye(m, dtype=A.dtype)
+    Z = math.zeros((m, m), dtype=A.dtype)
+    X = math.block([[Z, I], [I, Z]])
+    M = math.gather(math.gather(A, idx, axis=-1), idx, axis=-2) + X * measure
+    bM = math.gather(b, idx, axis=-1)
+    cpart1 = math.sqrt(math.cast((-1) ** m / math.det(M), "complex128"))
```

Comment:
I'm remembering having to introduce some logic for handling the case when `math.det(M)` is zero (see complex_gaussian_integral). E.g. this was an issue when calling (I believe) .normalize on QuadratureEigenstate. Should we have that here?

Reply:
I'm not sure. What were you doing in those situations?